apollo-router/src/configuration/cors.rs

```diff
@@ -84,90 +154,338 @@ impl Cors {
+                    return Some(origin_config);
+                }
+            }
+
+            // Check regex matches
+            if !origin_config.match_origins.is_empty() {
+                for regex in &origin_config.match_origins {
+                    if regex.is_match(origin_str) {
+                        return Some(origin_config);
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
+    /// Add CORS headers to the response
+    fn add_cors_headers<ResBody>(
+        response: &mut Response<ResBody>,
+        config: &Cors,
+        origin_config: &Option<&OriginConfig>,
+        origin: &Option<http::HeaderValue>,
+        is_preflight: bool,
+        request_headers: Option<http::HeaderValue>,
+    ) {
+        // Determine which configuration to use (origin-specific or global)
+        let allow_credentials = origin_config
+            .and_then(|oc| oc.allow_credentials)
+            .unwrap_or(config.allow_credentials);
+
+        let allow_headers = origin_config
+            .and_then(|oc| {
+                if oc.allow_headers.is_empty() {
+                    None
+                } else {
+                    Some(&oc.allow_headers)
+                }
+            })
+            .unwrap_or(&config.allow_headers);
+
+        let expose_headers = origin_config
+            .map(|oc| oc.expose_headers.as_ref())
+            .or(config.expose_headers.as_ref());
+
+        let methods = origin_config
+            .and_then(|oc| {
+                if oc.methods.is_empty() {
+                    None
+                } else {
+                    Some(&oc.methods)
+                }
+            })
+            .unwrap_or(&config.methods);
+
+        let max_age = config.max_age;
+
+        // Set Access-Control-Allow-Origin
+        if let Some(origin) = origin {
+            if config.allow_any_origin {
+                response.headers_mut().insert(
+                    ACCESS_CONTROL_ALLOW_ORIGIN,
+                    http::HeaderValue::from_static("*"),
+                );
+            } else if origin_config.is_some() {
+                // Only set the header if we found a matching origin configuration
+                response
+                    .headers_mut()
+                    .insert(ACCESS_CONTROL_ALLOW_ORIGIN, origin.clone());
+            }
+            // If no matching origin config found, don't set the header (origin will be rejected)
+        }
+
+        // Set Access-Control-Allow-Credentials
+        if allow_credentials {
+            response.headers_mut().insert(
+                ACCESS_CONTROL_ALLOW_CREDENTIALS,
+                http::HeaderValue::from_static("true"),
+            );
+        }
+
+        // Set Access-Control-Allow-Headers
+        if !allow_headers.is_empty() {
+            for header in allow_headers {
+                // Use the header value as-is to match test expectations
+                response.headers_mut().append(
```

Comment:
My Rust is weak as well as I dont know these basic http classes well.  Isn't there a  optimization available here on precomputing the headervalue and doing an insert once instead of append?  less lookups/memory allocations